[quantization] Allow set for qconfig for dynamic_quantize

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27184 [quantization] Fix reprs for _intrinsic modules
* **#27181 [quantization] Allow set for qconfig for dynamic_quantize**

Differential Revision: [D17717482](https://our.internmc.facebook.com/intern/diff/D17717482)

Commit:
[quantization] Allow set for qconfig for dynamic_quantize

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -385,6 +394,9 @@ def checkQuantized(model):
         model = quantize_dynamic(NestedModel().eval(), qconfig_dict)
         checkQuantized(model)
 
+        model = quantize_dynamic(NestedModel().eval(), set(['fc3', 'sub2.fc1']))
```
nit: `{'a', 'b'}` syntax also works

dzhulgakov commented on torch/quantization/quantize.py:
```diff
@@ -270,48 +271,54 @@ def quantize(model, run_fn, run_args, mapping=DEFAULT_MODULE_MAPPING, inplace=Fa
+              module unless qconfig for the submodules are specified (when the
+              submodule already has qconfig attribute). Entries in the dictionary
+              need to be QConfigDynamic instances.
+            * A set of types and/or submodule names to apply dynamic quantization to
```
add "in which case `dtype` is used"

dzhulgakov commented on torch/quantization/quantize.py:
```diff
@@ -270,48 +271,54 @@ def quantize(model, run_fn, run_args, mapping=DEFAULT_MODULE_MAPPING, inplace=Fa
             raise ValueError(
                 "Don't know how to quantize with default settings for {}. Provide full qconfig please".format(dtype))
+    elif isinstance(qconfig, set):
+        qconfig = dict(zip(qconfig, itertools.repeat(default_dynamic_qconfig)))
```
even though float16 is not official yet, you should use dtype here to choose default_dynamic_qconfig or float16_dynamic_qconfig

dzhulgakov commented on torch/quantization/quantize.py:
```diff
@@ -270,48 +271,54 @@ def quantize(model, run_fn, run_args, mapping=DEFAULT_MODULE_MAPPING, inplace=Fa
+    elif isinstance(qconfig, set):
+        qconfig = dict(zip(qconfig, itertools.repeat(default_dynamic_qconfig)))
+
+    print(qconfig)
```
remove

Commit:
Update on "[quantization] Allow set for qconfig for dynamic_quantize"

Commit:
Update on "[quantization] Allow set for qconfig for dynamic_quantize"

Commit:
Update on "[quantization] Allow set for qconfig for dynamic_quantize"

dzhulgakov reviewed (approved):
thanks!

Commit:
Update on "[quantization] Allow set for qconfig for dynamic_quantize"


Differential Revision: [D17717482](https://our.internmc.facebook.com/intern/diff/D17717482)

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -270,48 +271,59 @@ def quantize(model, run_fn, run_args, mapping=DEFAULT_MODULE_MAPPING, inplace=Fa
-            module unless qconfig for the submodules are specified (when the
-            submodule already has qconfig attribute). Entries in the dictionary
-            need to be QConfigDynamic instances.
+        qconfig: Either:
```
Lets call this something else. qconfig already refers to observer/fake-quant configuration and having this mean something else is confusing.

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -270,48 +271,59 @@ def quantize(model, run_fn, run_args, mapping=DEFAULT_MODULE_MAPPING, inplace=Fa
     return model
 
-def quantize_dynamic(model, qconfig_dict=None, dtype=torch.qint8, mapping=DEFAULT_DYNAMIC_MODULE_MAPPING, inplace=False):
+def quantize_dynamic(model, qconfig=None, dtype=torch.qint8, mapping=DEFAULT_DYNAMIC_MODULE_MAPPING, inplace=False):
```
Is mapping used anymore?

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -270,48 +271,59 @@ def quantize(model, run_fn, run_args, mapping=DEFAULT_MODULE_MAPPING, inplace=Fa
     return model
 
-def quantize_dynamic(model, qconfig_dict=None, dtype=torch.qint8, mapping=DEFAULT_DYNAMIC_MODULE_MAPPING, inplace=False):
+def quantize_dynamic(model, qconfig=None, dtype=torch.qint8, mapping=DEFAULT_DYNAMIC_MODULE_MAPPING, inplace=False):
```
Why not have an API with simpler meaning for qconfig and refactor the qconfig_dict creation logic into a seperate function? Currently qconfig can be a set of module names or just a dtype or a dictionary, which is confusing.


```
quantize_dynamic(model, torch.quantization.get_dynamic_qconfig_dict(quantization_set, dtype), mapping, inplace)
```
```python
# function to take in quantization set and dtype info and return a qconfig_dict
def get_dynamic_qconfig_dict(quantization_set=None, dtype=torch.qint8):
   if quantization_set is None:      
       if dtype == torch.qint8:
             qconfig = {
                nn.Linear : default_dynamic_qconfig,
                nn.LSTM : default_dynamic_qconfig,
            }
        elif dtype == torch.float16:
             qconfig = {
                # TODO: uncomment when float16 Linear support is added
                # nn.Linear : default_dynamic_qconfig,
                nn.LSTM : float16_dynamic_qconfig,
            }
        else:
            raise ValueError(
                "Don't know how to quantize with default settings for {}. Provide full qconfig please".format(dtype))
    elif isinstance(quantization_set, set):
        if dtype is torch.qint8:
            default_qconfig = default_dynamic_qconfig
        elif dtype is torch.float16:
            default_qconfig = float16_dynamic_qconfig
        else:
            raise RuntimeError('Unknown dtype specified for quantize_dynamic: ', str(dtype))
        qconfig = dict(zip(qconfig, itertools.repeat(default_qconfig)))
   return qconfig
```

jamesr66a replied:
That looks way more complicated from an end user perspective...?

jamesr66a replied:
The purpose of this PR is to *reduce* the amount of boilerplate the user has to write, what you've proposed does the exact opposite

raghuramank100 replied:
I think it is clearer from a end user perspective to just say:
```
quantize_dynamic(model, torch.quantization.get_dynamic_qconfig_dict(quantization_set, dtype)..)
```
rather than have qconfig_spec be a set or modules or dtypes or a dict. However, if you and @dzhulgakov  feel that this is ok, I will not object further.

raghuramank100 reviewed (dismissed):
Can we simplify this API, the current design overloads qconfig to mean too many things.
